packages/knip/src/util/modules.ts

```diff
@@ -64,6 +64,7 @@ export const sanitizeSpecifier = (specifier: string) => {
     }
   }
   s = s.substring(0, end);
-  if (s.includes(':') && !s.includes('/')) s = s.split(':')[0];
+  if (s.startsWith('astro:')) s = 'astro';
+  else if (s.includes(':') && !s.includes('/')) s = s.split(':')[0];
```

Comment:
```suggestion
  if (/[^/]+:/.test(s)) s = s.split(':')[0];
```

So if there's no slash before the colon, return the part before the colon. Otherwise it could be a path param (`/:param`). Node built-ins like `node:fs` are caught with `isBuiltin`. Not that it's good or pretty or whatever, but here we are.. WDYT?

Reply:
I came from a regex solution and wanted to fix a bug, increase maintainability and also improve performance (this function is used _a lot_ in pretty hot code path - maybe results should even be cached).

So in that sense, your initial solution was good and my proposal was actually not that great in terms of performance: re-introduced a regex while there's already an iteration.

So now that we're on it.. what do you think of maximizing performance with a single iteration, ideally no regexes, while fixing your bug too? Do you maybe see a way to go about this?

Reply:
Thoughts on 

```ts
const CHAR_EXCLAMATION = 33; // '!'
const CHAR_DASH = 45; // '-'
const CHAR_SLASH = 47; // '/'
const CHAR_COLON = 58; // ':'
const CHAR_HASH = 35; // '#'
const CHAR_QUESTION = 63; // '?'

// Strip `?search` and other proprietary directives from the specifier (e.g. https://webpack.js.org/concepts/loaders/)
export const sanitizeSpecifier = (specifier: string) => {
  if (isBuiltin(specifier) || isAbsolute(specifier) || specifier.startsWith(PROTOCOL_VIRTUAL)) return specifier;
  const len = specifier.length;
  let start = 0;
  let end = len;
  let colon = -1;
  let sawSlash = false;
  for (let i = 0; i < len; i++) {
    const ch = specifier.charCodeAt(i);
    if (i === start && (ch === CHAR_EXCLAMATION || ch === CHAR_DASH)) {
      start++;
      continue;
    }
    if (ch === CHAR_SLASH && colon === -1) {
      sawSlash = true;
    }
    if (colon === -1 && ch === CHAR_COLON && !sawSlash) {
      colon = i;
    }
    if (ch === CHAR_EXCLAMATION || ch === CHAR_QUESTION || (ch === CHAR_HASH && i > start)) {
      end = i;
      break;
    }
  }
  return colon !== -1 && colon < end ? specifier.slice(start, colon) : specifier.slice(start, end);
};
```
- Single-pass algorithm instead of multiple passes (while loop, substring, for loop, regex test, split)
- Uses `charCodeAt()` for numeric comparisons instead of string character comparisons
- Caches string length to avoid repeated `.length` lookups
- Extracts substrings only once at the end instead of multiple intermediate `substring()` calls
- Eliminates regex operations - replaces `/[^/]+:/.test()` and `.split(':')` with inline tracking

We can also add simple unbounded `Map` based cache if we want.

Reply:
Wonderful! Ticks all the boxes 😃 If this passes the tests I'm more than happy to merge.

Let's defer caching for now. It could be beneficial, but at the cost of a bit of memory and I'm not yet sure how it'd stack up against other, perhaps more impactful areas such as creating TS programs and disk reads.